### DISCUSSION TITLE
Running CLI with a conflicting service on port 9000

### DISCUSSION THREAD

**Discussion by mattyindustries** [GoogleChromeLabs/ps-analysis-tool#622] [Support Forum] [answered]
**Describe the bug**
I'm trying to run the PSAT CLI but when I run it I get
```
npm run cli -- -s https://domain-aaa.com
Error: Report server port (9000) already in use. You might be already running CLI
```

I have an existing process that I can't change already running on port 9000.

I wasn't able to see an option in the documentation to be able to launch using a port other than 9000.
**Expected behavior**
An option to select another port to run than 900

**Comment by milindmore22** [comment]
Hello @mattyindustries 

Thanks for the great suggestion! We’re definitely considering adding a CLI option for users to control the port.

In the meantime, here are some ways to work around the port 9000 error:

- **Another application using port 9000:** If another program is already using port 9000, you’ll need to close it to free up the port.
- **Running multiple PSAT CLI instances:** If you’re trying to run two PSAT CLI sessions at once, stop the first one (press Command+C) before starting the second.

**Comment by mattyindustries** [comment]
Thanks @milindmore22 

In this case I'm working on a corporate laptop with a Zscalar proxy service running on port 9000. I'm unable to change this or stop the service so until I can select a different port I can't use this tool.

**Comment by milindmore22** [comment]
Sorry for the inconvenience, but you will need to do a small code change in 2 files and re-build CLI to make it work on your laptop with custom ports.

> [!NOTE]
> While adding a custom port, please consider using a 4 digit port, as most 2-3 digit ports are often used by admin/root access on the system.


```ps-analysis-tool/packages/cli/src/index.ts```  On the line `78`, `82` and `102` change port 9000 to 9090 

```ps-analysis-tool/packages/cli-dashboard/webpack.config.cjs``` on the line `35` change port 9000 to 9090

Once you save changes, run the following command

```
npm i && npm run cli:build 
```
Now you can run and check PSAT CLI for analysis
```
npm run cli -- -u https://yourwebsite.com/
```

Please let me know if you need further assistance.

**Comment by mattyindustries** [comment]
Thanks @milindmore22 

That gets me past the conflicted port but then I'm having an issue where the -u isn't being understood.

```
npm run cli -- -u https://www.google.com.au/

> ps-analysis-tool@0.6.0 cli
> node dist/cli/index.js https://www.google.com.au/

Please provide one and only one of the following
        a) URL of a site (-u or --url)
        b) URL of a sitemap (-s or --sitemap-url)
        c) Path to a CSV file (-c or --csv-path)
        d) Path to an XML file (-p or --sitemap-path)
```

I'm not quite sure why that is happening

**Comment by milindmore22** [comment]
@mattyindustries Thanks for letting us know about the issue!

We reviewed the output you provided and noticed that the logged command was missing the `-u` flag.
The correct command log should be:

`> node dist/cli/index.js -u https://www.google.com.au/`

While we were able to run the command successfully on our end, we suspect the difference might be due to varying character encoding on your system.

E.g.: if we try putting `option + space` between -- and -u.

`npm run cli --<option + space>-u https://www.google.com.au/`

We can replicate your issue.

**Comment by mattyindustries** [comment]
Thanks @milindmore22.

I'm running this on a Windows 11 laptop using Microsoft Terminal.

Do you want me to log a seperate bug for this issue?

**Comment by milindmore22** [comment]
@mattyindustries Thank you for letting us know about your OS, When we tried on Windows 11 (with port change) we were able to run commands successfully.

![image](https://github.com/GoogleChromeLabs/ps-analysis-tool/assets/1132541/c51abcb0-3dde-4174-9037-43f005e1130c)

We have also logged an issue https://github.com/GoogleChromeLabs/ps-analysis-tool/issues/574 to validate CLI options and URL in CLI command, so it will show a warning when user enters wrong input.

I will recommend re-clone the repo and re-try it, also check on different machines that have non-restrictive environment (non-corporate laptop).

**Comment by milindmore22** [comment, marked as answer]
Hey @mattyindustries!

We're thrilled to announce the release of version **0.7.0** this update brings you more control over your application with a new CLI option for choosing a custom port to avoid conflicts.

To get started:

1.  Update your local copy by running `git pull` in your terminal.
2. Rebuild the CLI with `npm i` then `npm run cli:build`

Now you can use the new `-po` or `--port` flag

For example, to run the app on port 9009 and access BBC News, you'd use:

`npm run cli -- -po 9009 -u https://bbc.com`

Please check our updated [wiki](https://github.com/GoogleChromeLabs/ps-analysis-tool/wiki/PSAT-Command-Line-Interface#cli-options) for detailed information. 

I will be converting this issue to support forum discussion, For future reference, please [create support discussion](https://github.com/GoogleChromeLabs/ps-analysis-tool/discussions/new?category=support-forum) if you have any question about the PSAT.

**Comment by milindmore22** [comment]
Hey @mattyindustries 

I noticed your previous comment (https://github.com/GoogleChromeLabs/ps-analysis-tool/issues/574#issuecomment-2065536950) on GitHub https://github.com/sourcegraph/sourcegraph/pulls #issuecomment-2065536950.  To keep all our communication organized, I'd like to continue this discussion here.

> hey @milindmore22
> 
> I'm testing the recent update to support port options from https://github.com/GoogleChromeLabs/ps-analysis-tool/issues/573 and I'm still having this issue with the CLI.
> 
> npm run cli -- -po 10000 -u https://www.bbc.com
> 
> > ps-analysis-tool@0.7.0 cli
> > node dist/cli/index.js 10000 https://www.bbc.com
> 
> Please provide one and only one of the following
>         a) URL of a site (-u or --url)
>         b) URL of a sitemap (-s or --sitemap-url)
>         c) Path to a CSV file (-c or --csv-path)
>         d) Path to an XML file (-p or --sitemap-path)
> My Windows 11 machine seems to be not properly handling the -po part

The issue with missing CLI options is consistent on your machine, can you please run the `node -v` command on your terminal and send the node version you are using?

This information will help us investigate further.

**Comment by milindmore22** [comment]
Hello @mattyindustries,

I'm thrilled to announce the release of PSAT version **0.9.0** https://github.com/GoogleChromeLabs/ps-analysis-tool/discussions/732!

This new version brings a significant improvement: PSAT reports are now generated as HTML files. This eliminates the need for ports, preventing conflicts and making it easier to use PSAT.

To get started with PSAT v0.9.0:

1. Pull the latest changes: `git pull`

2. Rebuild the CLI:
   - Run `npm ci` to install dependencies.
   - Then, run `npm run build:all` to build the CLI.

3. Generate a report:
   Once the build is complete, run `npm run cli -- -u https://example.com`. This will create an HTML report for your website that you can view in any browser.

For more details on using the PSAT CLI, refer to our updated [wiki page](https://github.com/GoogleChromeLabs/ps-analysis-tool/wiki/PSAT-Command-Line-Interface).

If you have any questions or require further assistance, please don't hesitate to ask us on [support forums](https://github.com/GoogleChromeLabs/ps-analysis-tool/discussions/new?category=support-forum)
